src/Mafia/Install.hs

```diff
@@ -163,50 +181,96 @@ link w db env p@(Package (PackageRef pid _ _) _ _) = do
-
-        -- create symlinks to the relevant package .conf files in the
-        -- package-db, then call recache so that ghc is aware of them.
-        mapM_ (link w db env) (Set.toList (transitiveOfPackages deps))
-        Hush <- sbcabal "sandbox" ["hc-pkg", "recache"]
-
-        let constraints = concatMap (\c -> ["--constraint", c]) (constraintsOfPackage p)
-        Pass <- sbcabal "install" $ [ "--ghc-options=-j" <> T.pack (show w)
-                                    , "--ghc-options=-fprof-auto-exported"
-                                    , "--enable-library-profiling"
-                                    , "--enable-documentation"
-                                    , "--haddock-hoogle"
-                                    , "--haddock-hyperlink-source"
-                                    , "--max-backjumps=0"
-                                    , renderPackageId pid
-                                    ] <> constraints
-
-        Out out <- sbcabal "sandbox" ["hc-pkg", "--", "describe", renderPackageId pid]
-
-        writeUtf8 packageCfg out
+install :: NumWorkers -> PackageEnv -> Flavour -> Package -> EitherT InstallError IO ()
+install w penv flavour p@(Package (PackageRef pid _ _) deps _) = do
+  -- install package dependencies
+  mapM_ (install w penv flavour) (Set.toList (transitiveOfPackages deps))
+
+  -- the vanilla flavour must always be installed first:
+  --  + it creates and sets up the package's sandbox
+  --  + profiling builds need the vanilla build for template haskell to run
+  --  + documentation builds need the vanilla build to harvest the .hi files
+  when (flavour /= Vanilla) $
+    install w penv Vanilla p
+
+  let fmark = packageFlavourMarker penv p flavour
+  unlessM (doesFileExist fmark) $ do
+    withPackageLock penv p flavour $ do
+      unlessM (doesFileExist fmark) $ do
+        when (flavour == Vanilla) $
+          createPackageSandbox penv p
+
+        let sbdir        = packageSandboxDir penv p
+            sbcfg        = packageSandboxConfig penv p
+            sbcabal x xs = firstEitherT InstallCabalError $ cabalFrom sbdir (Just sbcfg) x xs
+
+        liftIO . T.putStrLn $ "Building " <> renderHashId p <> renderFlavourSuffix flavour
+
+        Pass <- sbcabal "install" $
+          [ "--ghc-options=-j" <> T.pack (show w)
+          , "--max-backjumps=0"
+          , renderPackageId pid ] <>
+          flavourArgs flavour <>
+          concatMap (\c -> ["--constraint", c]) (constraintsOfPackage p)
+
+        when (flavour == Vanilla) $ do
+          Out out <- sbcabal "sandbox" ["hc-pkg", "--", "describe", renderPackageId pid]
+          writeUtf8 (packageConfig penv p) out
+
+        writeBytes fmark B.empty
+
+flavourArgs :: Flavour -> [Argument]
+flavourArgs = \case
+  Vanilla ->
+    []
+  Profiling ->
+    [ "--reinstall"
+    , "--ghc-options=-fprof-auto-exported"
+    , "--enable-library-profiling" ]
+  Documentation ->
```

Comment:
yeah i think we'd call that when someone asks for hoogle or for haddock docs at some point